File: doc-examples/azure-cosmos-example-java/src/main/java/example/PersonRepository.java

package example;

import io.micronaut.data.cosmos.annotation.CosmosRepository;
import io.micronaut.data.model.Page;
import io.micronaut.data.model.Pageable;
import io.micronaut.data.model.Sort;
import io.micronaut.data.repository.CrudRepository;
import io.micronaut.data.repository.jpa.JpaSpecificationExecutor;
import io.micronaut.data.repository.jpa.criteria.DeleteSpecification;
import io.micronaut.data.repository.jpa.criteria.PredicateSpecification;
import io.micronaut.data.repository.jpa.criteria.QuerySpecification;
import io.micronaut.data.repository.jpa.criteria.UpdateSpecification;

import java.util.List;
import java.util.Optional;

// tag::repository[]
@CosmosRepository
public interface PersonRepository extends CrudRepository<Person, String>, JpaSpecificationExecutor<Person> {
// end::repository[]

    // tag::find[]
    @Override
    Optional<Person> findOne(PredicateSpecification<Person> spec);

    @Override
    Optional<Person> findOne(QuerySpecification<Person> spec);

    @Override
    List<Person> findAll(PredicateSpecification<Person> spec);

    @Override
    List<Person> findAll(QuerySpecification<Person> spec);

    @Override
    List<Person> findAll(PredicateSpecification<Person> spec, Sort sort);

    @Override
    List<Person> findAll(QuerySpecification<Person> spec, Sort sort);

    @Override
    Page<Person> findAll(PredicateSpecification<Person> spec, Pageable pageable);

    @Override
    Page<Person> findAll(QuerySpecification<Person> spec, Pageable pageable);
    // end::find[]

    // tag::count[]
    @Override
    long count(PredicateSpecification<Person> spec);

    @Override
    long count(QuerySpecification<Person> spec);
    // end::count[]

    // tag::update[]
    @Override
    long updateAll(UpdateSpecification<Person> spec);
    // end::update[]

    // tag::delete[]
    @Override
    long deleteAll(PredicateSpecification<Person> spec);

    @Override
    long deleteAll(DeleteSpecification<Person> spec);
    // end::delete[]

    // tag::specifications[]
    // tag::allSpecifications[]
    class Specifications {

        static PredicateSpecification<Person> nameEquals(String name) {
            return (root, criteriaBuilder) -> criteriaBuilder.equal(root.get("name"), name);
        }

        static PredicateSpecification<Person> ageIsLessThan(int age) {
            return (root, criteriaBuilder) -> criteriaBuilder.lessThan(root.get("age"), age);
        }

        // end::specifications[]
        static UpdateSpecification<Person> setNewName(String newName) {
            return (root, query, criteriaBuilder) -> {
                // tag::setUpdate[]
                query.set(root.get("name"), newName);
                // end::setUpdate[]
                return null;
            };
        }
        // tag::specifications[]
    }
    // end::allSpecifications[]
    // end::specifications[]
// tag::repository[]
}
// end::repository[]

File: doc-examples/azure-cosmos-example-kotlin/src/main/kotlin/example/PersonRepository.kt
package example

import io.micronaut.data.cosmos.annotation.CosmosRepository
import io.micronaut.data.model.Page
import io.micronaut.data.model.Pageable
import io.micronaut.data.model.Sort
import io.micronaut.data.repository.CrudRepository
import io.micronaut.data.repository.jpa.JpaSpecificationExecutor
import io.micronaut.data.repository.jpa.criteria.DeleteSpecification
import io.micronaut.data.repository.jpa.criteria.PredicateSpecification
import io.micronaut.data.repository.jpa.criteria.QuerySpecification
import io.micronaut.data.repository.jpa.criteria.UpdateSpecification
import io.micronaut.data.runtime.criteria.get
import io.micronaut.data.runtime.criteria.query
import io.micronaut.data.runtime.criteria.update
import io.micronaut.data.runtime.criteria.where
import java.util.*

// tag::repository[]
@CosmosRepository
interface PersonRepository : CrudRepository<Person, String>, JpaSpecificationExecutor<Person> {
    // end::repository[]
    override
    // tag::find[]
    fun findOne(spec: PredicateSpecification<Person>?): Optional<Person>

    // end::find[]
    override
    // tag::find[]
    fun findOne(spec: QuerySpecification<Person>?): Optional<Person>

    // end::find[]
    override
    // tag::find[]
    fun findAll(spec: PredicateSpecification<Person>?): List<Person>

    // end::find[]
    override
    // tag::find[]
    fun findAll(spec: QuerySpecification<Person>?): List<Person>

    // end::find[]
    override
    // tag::find[]
    fun findAll(spec: PredicateSpecification<Person>?, sort: Sort): List<Person>

    // end::find[]
    override
    // tag::find[]
    fun findAll(spec: QuerySpecification<Person>?, sort: Sort): List<Person>

    // end::find[]
    override
    // tag::find[]
    fun findAll(spec: PredicateSpecification<Person>?, pageable: Pageable): Page<Person>

    // end::find[]
    override
    // tag::find[]
    fun findAll(spec: QuerySpecification<Person>?, pageable: Pageable): Page<Person>

    // end::find[]
    override
    // tag::count[]
    fun count(spec: PredicateSpecification<Person>?): Long

    // end::count[]
    override
    // tag::count[]
    fun count(spec: QuerySpecification<Person>?): Long

    // end::count[]
    override
    // tag::update[]
    fun updateAll(spec: UpdateSpecification<Person>?): Long

    // end::update[]
    override
    // tag::delete[]
    fun deleteAll(spec: PredicateSpecification<Person>?): Long

    // end::delete[]
    override
    // tag::delete[]
    fun deleteAll(spec: DeleteSpecification<Person>?): Long

    // end::delete[]
    // tag::specifications[]
    // tag::allSpecifications[]
    object Specifications {
        // tag::where[]
        fun nameEquals(name: String?) = where<Person> { root[Person::name] eq name }

        fun ageIsLessThan(age: Int) = where<Person> { root[Person::age] lt age }

        fun nameInList(names: List<String>) = where<Person> { root[Person::name] inList names }
        // end::where[]

        // tag::or[]
        fun nameOrAgeMatches(age: Int, name: String?) = where<Person> {
            or {
                root[Person::name] eq name
                root[Person::age] lt age
            }
        }

        fun nameAndAgeMatch(age: Int, name: String) = query<Person> {
            where {
                root[Person::name] eq name
                root[Person::age] lt age
            }
        }
        // end::or[]

        // end::specifications[]
        // tag::setUpdate[]
        fun updateName(newName: String, existingName: String) = update<Person> {
            set(Person::name, newName)
            where {
                root[Person::name] eq existingName
            }
        }
        // end::setUpdate[]

        // Different style using the criteria builder
        fun nameEquals2(name: String?) = PredicateSpecification { root, criteriaBuilder ->
            criteriaBuilder.equal(root[Person::name], name)
        }

        fun ageIsLessThan2(age: Int) = PredicateSpecification { root, criteriaBuilder ->
            criteriaBuilder.lessThan(root[Person::age], age)
        }

        fun setNewName2(newName: String) = UpdateSpecification { root, query, _ ->
            // tag::setUpdate[]
            query.set(root[Person::name], newName)
            // end::setUpdate[]
            null
        }

        // tag::specifications[]
    }
    // end::allSpecifications[]
    // end::specifications[]
    // tag::repository[]
}

